File: app/src/test/java/com/zwq65/unity/pattern/dynamic_proxy/LoggerHandler.java
package com.zwq65.unity.pattern.dynamic_proxy;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

/**
 * ================================================
 * <p>
 * Created by NIRVANA on 2017/11/15
 * Contact with <zwq651406441@gmail.com>
 * ================================================
 */
public class LoggerHandler implements InvocationHandler {

    private Object target;

    public LoggerHandler(Object target) {
        this.target = target;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        System.out.println("準備打印");
        Object response = method.invoke(target, args);
        System.out.println("結束打印");
        return response;
    }
}

File: app/src/test/java/com/zwq65/unity/pattern/dynamic_proxy/LoggerHandler.kt
package com.zwq65.unity.pattern.dynamic_proxy

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

/**
 * ================================================
 * <p>
 * Created by NIRVANA on 2017/11/15
 * Contact with <zwq651406441@gmail.com>
 * ================================================
 */
class LoggerHandler(private val target: Any) : InvocationHandler {

    @Throws(Throwable::class)
    override fun invoke(proxy: Any, method: Method, args: Array<Any>): Any {
        println("準備打印")
        val response = method.invoke(target, *args)
        println("結束打印")
        return response
    }
}
